File: find_transmission_lineages.pl
#!/usr/bin/perl

# Input: 1) --tree newick tree 2) --states fasta with states for all nodes (including leaves)
# Output format: list of nodes closest to root where probability of Russian state (2) turns to 1.0
# >Entry_node
# list of strains in the corresponding subtree (if entry_node is not a leaf)

use Getopt::Long;
use Bio::Phylo::IO;
use strict;

my $tree;
my $states;
my $output;

GetOptions (	
		'tree=s' => \$tree,
		'states=s' => \$states,
		'output=s' =>\$output,
		);
$| = 1;

print "Parsing tree.. \n";
my $st = time();
my $tree = parse_tree($tree);
my $end = time();
my $runtime = $end-$st;
print "Finished in $runtime sec \n";
my %states;
print "Parsing states.. \n";
$st = time();
open ST, "<$states" or die "cannot open states file $states $! \n";
while(<ST>){
	my ($nname, $state1_prob, $state2_prob) = split(/\s+/);
	$states{$nname} = $state2_prob; # probability of being russian
	print "nname $nname state $state2_prob\n";
}

close ST;
$end = time();
$runtime = $end-$st;
print "Finished in $runtime sec \n";

my %lineages; # lineages{entry_node_name} = @strains

print "Looking for lineages.. \n";
$st = time();
my @array;
my_visit_depth_first ($tree->get_root, \@array,  \&collect_lineage_members,\&not_exported, "");
$end = time();
$runtime = $end-$st;
print "Finished in $runtime sec \n";

foreach my $entry_nname(keys %lineages){
	print ">".$entry_nname."\n";
	print join("\n", @{$lineages{$entry_nname}})."\n";
}

sub collect_lineage_members {
	my $node = $_[0];
	my $entry_nname = $_[1];
	#print "Node ".$node->get_name()." Entry name $entry_nname\n";
	if ($entry_nname){
		if ($node->is_terminal){push $lineages{$entry_nname}, $node->get_name();}
	}
	else {
		#print "Probability of being russian is ".$states{$node->get_name()}."\n";
		if ($states{$node->get_name()} == 1 ){ # probability of being russian is 1
			$entry_nname = $node->get_name();
			print "Lineage started at $entry_nname\n";
			my @linlist;
			$lineages{$entry_nname} = \@linlist;
			if ($node->is_terminal){push $lineages{$entry_nname}, $node->get_name();}
		}
	}
	return $entry_nname;
} 

sub not_exported {
	my $node = $_[0];
	my $entry_nname = $_[1];
	if ($entry_nname && $states{$node->get_name()} == 0){return 0;} # already been to russia but now probability of being russian is 0
	else {return 1;}
}


sub parse_tree {
					my $tree_file = $_[0];
					open TREE, "<$tree_file" or die "Cannot open file ".$tree_file."\n";
					# get a newick string from some source
 					my $tree_string;
 					my $t;
 					while($t = <TREE>){
 						$t =~ s/\n$//;
 						$tree_string = $tree_string.$t;
 					}
 					 close TREE;

 					# Call class method parse from Bio::Phylo::IO
 					# note: newick parser returns 'Bio::Phylo::Forest'
                    # Call ->first to retrieve the first tree of the forest.
 					my $tree = Bio::Phylo::IO->parse(
  					  -string => $tree_string,
   					  -format => 'newick'
 					)->first;

 					return $tree;
	
} 

 sub my_visit_depth_first {
		my $node = $_[0];
		my @array = @{$_[1]};
		my $action_callback = $_[2];
		my $check_callback = $_[3];
		my $action_result = $_[4];
		push @array, $node;
		my $ok = &$check_callback($node, $action_result);
		#print "Processing node ".$node->get_name()."\n";
		#print "Is it ok? (not exported) $ok \n";
		if ($ok) {$action_result = &$action_callback($node, $action_result)}; # current $entry_nname
		if (! $node->is_terminal && $ok){
			my $i = 0;
			while($node->get_child($i)){
				@array = my_visit_depth_first($node->get_child($i), \@array, \&$action_callback, \&$check_callback, $action_result);
				$i++;
			}
		}	
		$node = pop @array;
		#print $node->get_name()."\n";
		return @array;
    }

File: find_transmission_lineages.py
#from Bio import Phylo
from ete3 import Tree
import time
import pandas as pd
import optparse


parser = optparse.OptionParser()
parser.add_option('-t', '--tree', help='tree file', type='str')
parser.add_option('-s', '--states', help='tab-delimited file with states for all nodes', type='str')
parser.add_option('-d', '--duplicates', help='file with duplicates (and cluster memebers)', type='str')
parser.add_option('-c', '--countries', help='file with ids and countries, output from meta_to_states.py', type='str')
parser.add_option('-p', '--print_broader_subtree', default=-1, help='return newick subtrees for each entry. Make N steps up the tree before extracting a subtree', type='int')
parser.add_option('-o', '--output', help='output', type='str')

options, args = parser.parse_args()

lineages = {} # lineages{entry_node_name} = @strains


def find_primary_lineages(node, array, entry_nname):
	array.append(node)
	exported = entry_nname is not None and states[node.name] == 0
	print("Processing node " + node.name + " with state 2 probability " + str(states[node.name]) + " and exproted = " + str(exported))
	#print("Is it exported? " + str(exported))
	#if len(lineages) > 0:
	#	print("Lineage len " + str(len(lineages)))
	if not exported:
		if entry_nname is not None:
			if node.is_leaf():
				lineages[entry_nname].append(node.name)
		else:
			if states[node.name] == 1:  # probability of being russian is 1
				entry_nname = node.name
				print("Lineage started at " + entry_nname)
				lineages[entry_nname] = []
				if node.is_leaf():
					lineages[entry_nname].append(node.name)
				#else:
					 #node.write(format=1, outfile=options.output + "_entry_" + node.name + ".newick")
	if not node.is_leaf() and not exported:
		for child in node.children:
			array = find_lineages(child, array, entry_nname)
	node = array.pop()
	return(array)


def find_lineages(node, array, entry_nname):
	array.append(node)
	exported = entry_nname is not None and states[node.name] == 0
	if exported:
		entry_nname = None
	print("Processing node " + node.name + " with state 2 probability " + str(states[node.name]) + " and exproted = " + str(exported))
	#print("Is it exported? " + str(exported))
	#if len(lineages) > 0:
	#	print("Lineage len " + str(len(lineages)))
	if not exported and entry_nname is not None:
		if node.is_leaf():
			lineages[entry_nname].append(node.name)
	if entry_nname is None:
		if states[node.name] == 1:  # probability of being russian is 1
			entry_nname = node.name
			print("Lineage started at " + entry_nname)
			lineages[entry_nname] = []
			if node.is_leaf():
				lineages[entry_nname].append(node.name)
				#else:
					 #node.write(format=1, outfile=options.output + "_entry_" + node.name + ".newick")
	if not node.is_leaf():
		for child in node.children:
			array = find_lineages(child, array, entry_nname)
	node = array.pop()
	return(array)


print("Parsing tree..")
now = time.time()
tree = Tree(options.tree, format = 1)
later = time.time()
diff = int(later - now)
print("Finished parsing in " + str(diff) + " seconds")


print("Parsing states..")
states = {}
with open(options.states) as st:
	for line in st:
		[nname, _, state2_prob] = line.split()
		states[nname] = float(state2_prob)  # probability of being russian
		print("nname " + nname + " state 2 probability " + str(state2_prob))


array = []
print("Looking for lineages.. ")
find_lineages(tree.get_tree_root(), array, None)

print("Parsing countries..")
countries = pd.read_csv(options.countries, sep="\t", names=['seq_id', 'state', 'region'])
countries_dict = dict(zip(countries['seq_id'], countries['region']))
print(dict(list(countries_dict.items())[0:5]))

print("Adding duplicates..")
duplicates = {}
with open(options.duplicates) as dup:
	for line in dup:
		splitter =  line.strip().split("\t")
		if len(splitter) >1:
			for value in lineages.values():
				[centroid, strain_list_str] = splitter
				if centroid in value:
					rus_strain_list = [s for s in strain_list_str.split(";") if s in countries_dict and countries_dict[s] == "Russia"]
					value.extend(rus_strain_list)



linsizes = {}
for entry_nname in lineages.keys():
	size = len(lineages[entry_nname])
	if not size in linsizes:
		linsizes[size] = []
	linsizes[size].append(entry_nname)

with open(options.output + "stats", "w") as out:
	for size in sorted(linsizes.keys()):
		out.write("Found " + str(len(linsizes[size])) + " lineages of size " + str(size) + ": "  + ",".join(linsizes[size]) + "\n")

with open(options.output, "w") as out:	
	for entry_nname in lineages.keys():
		out.write(entry_nname + "\t")
		out.write(";".join(lineages[entry_nname]) + "\n")

with open(options.output + ".entries", "w") as out:
	for entry_nname in lineages.keys():
		out.write(entry_nname + "\n")

steps = options.print_broader_subtree
if steps > -1:
	for entry_nname in lineages.keys():
		entry_node = tree.search_nodes(name=entry_nname)[0]
		if not entry_node.is_leaf():
			counter = 0
			printing_node = entry_node
			while counter < steps:
				printing_node = printing_node.up
				counter += 1
			printing_node.write(format=1, outfile=options.output + "_entry_" + entry_node.name + "_" + str(steps) + "_steps_up.newick")

